### DISCUSSION TITLE
How to prevent UI elements from overflowing the native window space?

### DISCUSSION THREAD

**Discussion by twokilohertz** [emilk/egui#4335] [Q&A]
I'm writing a native Windows application using egui, with `eframe::run_native()`, however, I'm having trouble with long text overflowing the width of the window. 

I have a grid which is populated with checkboxes which can potentially have very long strings of text associated with them. When these are added to the user interface, the text overflows the screen, expanding the `CentralPanel` causing everything to shift outside of the visible window until it's manually resized.

As a demonstration I've attached two images: the first which shows the UI as it is when the program is first launched & another image showing the UI shifting outside the native window when long text is added.

![Initial UI](https://github.com/emilk/egui/assets/72780006/6e6203f0-8a16-4cc6-95ba-1a920d9a57f3)
![UI after adding long text](https://github.com/emilk/egui/assets/72780006/b2f9d6ca-7ab0-4656-9559-d7a1dceb781c)

I'm sure there is a solution I am missing but I am new to egui and even after trawling through the docs, I cannot find a way to constrain everything within the bounds of the native window.

**Comment by thmxv** [comment]
For most widgets, you can make the layout wrap them. But for labels, that can be too long to even wrap, you can use [`truncate`](https://docs.rs/egui/latest/egui/widgets/struct.Label.html#method.truncate).

**Reply by twokilohertz** [reply]
I did consider using a wrapping layout but I think truncation is what would be preferable. However, as far as I understand, `.truncate(true)` is only available on the label widget. I'm wondering how complex of a task it would be to bring this behavious to other widgets with text (such as comboboxes & checkboxes).

**Reply by thmxv** [reply]
I usually put the `CheckBox` (with no text) followed by a `Label` inside a horizontal layout. For the `ComboBox` I do not know.

Edit: I do not know how to do it for the `ComboBox` but you can try setting the widht with [`width`](https://docs.rs/egui/latest/egui/containers/struct.ComboBox.html#method.width). If that does not work, you can either draw the `ComboBox` and its menu yourself using [`show_ui`](https://docs.rs/egui/latest/egui/containers/struct.ComboBox.html#method.show_ui). Or passing a (`Arc`) `Galley` instead of a string to the `Into<WidgetText>` parameter(s). I do not have enough experience to show you some code that does that but, I think, it sounds more complicated that it really is.

Edit2: I think the easiest solution is probably to use the `Galley` thing. Try constructing the `Label` (setting the `truncate` property) but do not display it, than use [`layout_in_ui`](https://docs.rs/egui/latest/egui/widgets/struct.Label.html#method.layout_in_ui) to build the `Galley` corresponding to it. (This probably need to be used togheter with the `show_ui` thingy)

**Comment by YgorSouza** [comment]
I think my answer in #2882 applies here as well:

```rs
impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        use eframe::egui::{
            text::{LayoutJob, TextWrapping},
            ComboBox, TextFormat, TextStyle,
        };
        egui::CentralPanel::default().show(ctx, |ui| {
            let text = "Sphinx of black quartz, judge my vow";
            let truncated = |s: &str, ui: &egui::Ui| {
                let mut job = LayoutJob::default();
                let format = TextFormat {
                    font_id: TextStyle::Button.resolve(ui.style()),
                    ..Default::default()
                };
                job.append(s, 0.0, format);
                job.wrap = TextWrapping {
                    max_rows: 1,
                    break_anywhere: true,
                    ..Default::default()
                };
                job
            };
            ui.group(|ui| {
                ui.horizontal(|ui| {
                    ui.label("Combo");
                    ComboBox::from_id_source("id")
                        .wrap(true)
                        .selected_text(truncated(text, ui))
                        .show_ui(ui, |_| {});
                });
                ui.checkbox(&mut true, truncated(text, ui));
            });
        });
    }
}
```

![image](https://github.com/emilk/egui/assets/43298013/740217a7-9848-4062-90d8-c9f14e15c900)

But again, I don't know if this is the best way, since it is a little verbose. But I guess you can just declare it as a helper function somewhere and use it wherever you need. Also, it doesn't work if you give the `ComboBox` a label, or put a label after it in the layout. The label has to come first, and then the selected text gets truncated to fit the rest of the space.

**Reply by thmxv** [reply]
Definitely a better answer than mine. If you need to place something at the right of the `ComboBox` or the truncated `Label` you can work around it but using a `right_to_left` layout. The important thing is that you put the truncated widget last in the code, but it can still be in between other widgets from a geometric point of view.
